A29-xds-tls-security.md

```diff
@@ -0,0 +1,422 @@
+**`DistributorWatcher`** of the plugin. The plugin typically creates a new thread and a
+timer to periodically fetch or mint a new certificate and send the certificate (and the key) to
+all its watchers. 
+
+### Bootstrap File Additions for Security
+
+The [bootstrap file][bootstrap-file] with [server side support][A36-xds-protocol] needs further
+changes as described here. The following snippet shows the additions for the `file_watcher` plugin
+that is already implemented in all gRPC languages.
+```
+{
+  // "certificate_providers" contains configurations for all supported plugins 
+  "certificate_providers": {
+    "google_cloud_private_spiffe": { // certificate_provider_instance name
+      "plugin_name": "file_watcher", // name of the plugin
+      "config": {                    // config to be supplied to the plugin
+        "certificate_file": "/var/run/gke-spiffe/certs/certificates.pem",
+        "private_key_file": "/var/run/gke-spiffe/certs/private_key.pem",
+        "ca_certificate_file": "/var/run/gke-spiffe/certs/ca_certificates.pem",
+        "refresh_interval": "60s"
+      }
+    }
+  }
+}
+```
+
+`"certificate_providers"` is a new top level field whose value is a map (a JSON object).
+The key in this map is the certificate_provider_instance name and the value is a JSON object
+having exactly 2 fields:
+
+* `"plugin_name"` which is the name of the plugin (a string value), and
+* `"config"` which is the configuration for the plugin. The value of `"config"` is a JSON object
+whose schema is defined by that plugin.
+
+In the above example, the config for the file_watcher plugin contains 4 fields: the first 3
+fields are the file paths for the identity certificate, private key and the root (or CA)
+certificate respectively. The last field is the certificate refresh interval i.e. the interval
+to be used by the plugin to monitor the file paths to refresh the certificates.
+
+[bootstrap-file]: A27-xds-global-load-balancing.md#xdsclient-and-bootstrap-file
+
+#### Use of Bootstrap File by the Certificate Provider Plugin Framework
+
+The xDS control plane specifies a certificate_provider_instance name as described above in
+[CommonTlsContext Processing](#commontlscontext-processing). For this example, let's assume
+it uses `"google_cloud_private_spiffe"` as the `"instance_name"` in
+`tls_certificate_certificate_provider_instance`. gRPC looks up
+`"google_cloud_private_spiffe"` in the `"certificate_providers"` map loaded from the
+bootstrap file and use the JSON object value to extract the plugin name (value of
+`"plugin_name"` as a string) and configuration (value of `"config"` as a JSON object).
+It passes on these values to the `CertificateProviderStore` to retrieve
+(after instantiating if necessary) the associated `CertificateProvider` plugin.
+The caller of the framework registers a watcher with the plugin to receive
+certificate updates as described below.
+
+### Implementing Security in the xDS Flow
```

Comment:
I don't think we really need to rehash the overall flow here; that should already be covered earlier in the document.  Instead, I think this section should be converted into an "Implementation Details" section, which includes things like the internals of the CertificateProvider API (e.g., flow from CertificateProvider to distributor to TLS creds) or how data flows through LB policies in various languages.

Reply:
Okay, good point and I will convert this into an "Implementation Details" section. Regarding your suggestion

> ...or how data flows through LB policies in various languages.

I thought it would be better to leave out language specific implementation details from this gRFC similar to how other gRFCs have done. What do you think?

Reply:
It's kind of a gray area.  I think that for a gRFC in the `A` category (as opposed to those in the `L` category), the overall design and functional specification needs to be described in a language-agnostic way.  However, it's also fine to include some language-specific details to help implementors understand how things need to work, as long as those details are not the core of the document.

In this case, I do think it would be useful to include a high-level description of how data flows through the LB policies.  The comments you have earlier in the doc about how C-core uses channel args and Java uses an `SslContextProvider` are appropriate to mention as part of that.